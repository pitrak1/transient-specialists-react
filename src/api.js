import axios from 'axios'
import utils from './utils'

const attachInterceptors = () => {
  axios.interceptors.response.use(
    function(response) {
      return response.data.statusCode === 500
        ? Promise.reject(response)
        : response
    },
    function(error) {
      return Promise.reject(error)
    },
  )
}

const translateError = error => {
  if (error.includes('duplicate key value violates unique constraint')) {
    if (error.includes('idx_type_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_oem_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_model_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_equipment_serial_number')) {
      return 'Serial Number must be unique (case insensitive)'
    }
  }

  if (error.includes('violates foreign key constraint')) {
    if (error.includes('equipments_model_id_fkey')) {
      return 'Model cannot be deleted because it has related equipment'
    }

    if (error.includes('equipments_type_id_fkey')) {
      return 'Type cannot be deleted because it has related equipment'
    }

    if (error.includes('models_oem_id_fkey')) {
      return 'OEM cannot be deleted because it has related models'
    }
  }

  return error
}

const getIndex = (resource, options, success, failure) => {
  attachInterceptors()
  const { ascending, page, perPage, searchValue, sortBy } = options
  return axios
    .get(
      `${
        process.env.LAMBDA_ENDPOINT
      }${resource}?ascending=${ascending}&page=${page}&perPage=${perPage}&searchValue=${searchValue}&sortBy=${utils.pascalToSnake(
        sortBy,
      )}`,
    )
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

const getShow = (resource, id, options, success, failure) => {
  attachInterceptors()
  const { ascending, page, perPage, sortBy } = options
  return axios
    .get(
      `${
        process.env.LAMBDA_ENDPOINT
      }${resource}?id=${id}&ascending=${ascending}&page=${page}&perPage=${perPage}&sortBy=${utils.pascalToSnake(
        sortBy,
      )}`,
    )
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

const getNew = (resource, success, failure) => {
  attachInterceptors()
  return axios
    .get(`${process.env.LAMBDA_ENDPOINT}${resource}?new=true`)
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

export const postCreate = (resource, data, success, failure) => {
  attachInterceptors()
  return axios
    .post(`${process.env.LAMBDA_ENDPOINT}${resource}`, data)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export const patchUpdate = (resource, data, success, failure) => {
  attachInterceptors()
  return axios
    .patch(`${process.env.LAMBDA_ENDPOINT}${resource}`, data)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export const deleteDestroy = (resource, id, success, failure) => {
  attachInterceptors()
  return axios
    .delete(`${process.env.LAMBDA_ENDPOINT}${resource}?id=${id}`)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export default {
  getIndex,
  getShow,
  getNew,
  postCreate,
  patchUpdate,
  deleteDestroy,
}
