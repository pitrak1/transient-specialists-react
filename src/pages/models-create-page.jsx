import React from 'react'
import { Button, CircularProgress, Grid, Typography } from '@material-ui/core'
import FormTextField from '../components/form-text-field.jsx'
import FormSelect from '../components/form-select.jsx'
import api from '../api.js'
import { withRouter } from 'react-router'

export class ModelsCreatePage extends React.Component {
  state = {
    alert: null,
    data: {},
    error: null,
    loading: true,
    name: '',
    nameValid: false,
    oemId: 0,
    oemIdValid: false,
  }

  componentDidMount() {
    api.getNew(
      'models',
      result => {
        this.setState({ loading: false, data: result })
      },
      error => {
        this.setState({ loading: false, error })
      },
    )
  }

  handleChange = (identifier, value, valid) => {
    const state = {}
    state[identifier] = value
    state[`${identifier}Valid`] = valid
    this.setState(state)
  }

  handleClick = () => {
    this.setState({ loading: true, alert: null })
    api.postCreate(
      'models',
      { name: this.state.name, oemId: this.state.oemId },
      () => {
        this.props.history.push('/models')
      },
      error => {
        this.setState({ loading: false, alert: error })
      },
    )
  }

  render() {
    if (this.state.loading) {
      return <CircularProgress size={120} />
    }

    if (this.state.error) {
      return <div>{this.state.error}</div>
    }

    return (
      <Grid container>
        {this.state.alert && (
          <Grid item xs={12}>
            {this.state.alert}
          </Grid>
        )}
        <Grid item xs={12}>
          <Typography variant='h6'>Add Model</Typography>
        </Grid>
        <Grid item xs={12}>
          <FormTextField
            identifier='name'
            label='Name'
            onChange={this.handleChange}
            required={true}
            value={this.state.name}
          />
        </Grid>
        <Grid item xs={12}>
          <FormSelect
            defaultOptionLabel='Select an OEM'
            disabled={false}
            identifier={'oemId'}
            label='OEM'
            onChange={this.handleChange}
            options={this.state.data.oems}
            required={true}
            value={this.state.oemId}
          />
        </Grid>
        <Grid item xs={12}>
          <Button
            disabled={!(this.state.nameValid && this.state.oemIdValid)}
            onClick={this.handleClick}
          >
            Submit
          </Button>
        </Grid>
      </Grid>
    )
  }
}

export default withRouter(ModelsCreatePage)
