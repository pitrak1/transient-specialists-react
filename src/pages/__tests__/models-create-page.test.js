import React from 'react'
import { ModelsCreatePage } from '../models-create-page'
import { shallow } from 'enzyme'
import sinon from 'sinon'
import api from '../../api.js'

describe('ModelsCreatePage', () => {
  const data = {
    oems: [
      { id: 1, name: 'OEM 1' },
      { id: 2, name: 'OEM 2' },
      { id: 3, name: 'OEM 3' },
    ],
  }
  it('renders spinner if not yet loaded', () => {
    const node = shallow(<ModelsCreatePage />)
    expect(node).toMatchSnapshot()
  })

  it('renders an error if load fails', () => {
    api.getNew = sinon.stub().callsArgWith(2, 'some error')
    const node = shallow(<ModelsCreatePage />)
    expect(node).toMatchSnapshot()
  })

  it('renders form if load succeeds', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    const node = shallow(<ModelsCreatePage />)
    expect(node).toMatchSnapshot()
  })

  it('enables submit if fields are valid', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    const node = shallow(<ModelsCreatePage />)
    node.instance().handleChange('name', 'Model 1', true)
    node.instance().handleChange('oemId', 1, true)
    expect(node).toMatchSnapshot()
  })

  it('redirects to models if submit succeeds', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    api.postCreate = sinon.stub().callsArg(2)
    const push = sinon.stub()
    const node = shallow(<ModelsCreatePage history={{ push }} />)
    node.instance().handleClick()
    expect(push.firstCall.args[0]).toBe('/models')
  })

  it('displays alert if submit fails', () => {
    api.getNew = sinon.stub().callsArgWith(1, data)
    api.postCreate = sinon.stub().callsArgWith(3, 'some error')
    const node = shallow(<ModelsCreatePage />)
    node.instance().handleClick()
    expect(node).toMatchSnapshot()
  })
})
