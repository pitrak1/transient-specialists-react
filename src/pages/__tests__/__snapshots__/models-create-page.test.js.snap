// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelsCreatePage displays alert if submit fails 1`] = `
<div>
  <Alert
    liveRegionPoliteness="assertive"
    margin="x-small 0"
    open={true}
    screenReaderOnly={false}
    timeout={0}
    transition="fade"
    variant="error"
  >
    some error
  </Alert>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    Add Model
  </Heading>
  <FormTextInput
    identifier="name"
    label="Name"
    onChange={[Function]}
    required={true}
    value=""
  />
  <FormSelect
    defaultOptionLabel="Select an OEM"
    disabled={false}
    identifier="oemId"
    label="OEM"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 1,
          "name": "OEM 1",
        },
        Object {
          "id": 2,
          "name": "OEM 2",
        },
        Object {
          "id": 3,
          "name": "OEM 3",
        },
      ]
    }
    required={true}
    value={null}
  />
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    disabled={true}
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Submit
  </Button>
</div>
`;

exports[`ModelsCreatePage enables submit if fields are valid 1`] = `
<div>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    Add Model
  </Heading>
  <FormTextInput
    identifier="name"
    label="Name"
    onChange={[Function]}
    required={true}
    value="Model 1"
  />
  <FormSelect
    defaultOptionLabel="Select an OEM"
    disabled={false}
    identifier="oemId"
    label="OEM"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 1,
          "name": "OEM 1",
        },
        Object {
          "id": 2,
          "name": "OEM 2",
        },
        Object {
          "id": 3,
          "name": "OEM 3",
        },
      ]
    }
    required={true}
    value={1}
  />
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    disabled={false}
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Submit
  </Button>
</div>
`;

exports[`ModelsCreatePage renders an error if load fails 1`] = `
<Alert
  liveRegionPoliteness="assertive"
  margin="x-small 0"
  open={true}
  screenReaderOnly={false}
  timeout={0}
  transition="fade"
  variant="error"
>
  some error
</Alert>
`;

exports[`ModelsCreatePage renders form if load succeeds 1`] = `
<div>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    Add Model
  </Heading>
  <FormTextInput
    identifier="name"
    label="Name"
    onChange={[Function]}
    required={true}
    value=""
  />
  <FormSelect
    defaultOptionLabel="Select an OEM"
    disabled={false}
    identifier="oemId"
    label="OEM"
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": 1,
          "name": "OEM 1",
        },
        Object {
          "id": 2,
          "name": "OEM 2",
        },
        Object {
          "id": 3,
          "name": "OEM 3",
        },
      ]
    }
    required={true}
    value={null}
  />
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    disabled={true}
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Submit
  </Button>
</div>
`;

exports[`ModelsCreatePage renders spinner if not yet loaded 1`] = `
<Spinner
  as="div"
  renderTitle="Loading"
  size="large"
  variant="default"
/>
`;
